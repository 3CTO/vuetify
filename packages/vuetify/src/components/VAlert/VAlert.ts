// Styles
import '../../stylus/components/_alerts.styl'

// Extensions
import VSheet from '../VSheet'

// Components
import VBtn from '../VBtn'
import VIcon from '../VIcon'

// Mixins
import Toggleable from '../../mixins/toggleable'
import Transitionable from '../../mixins/transitionable'

// Types
import { VNode } from 'vue/types'
import mixins from '../../util/mixins'
import Themeable from '../../mixins/themeable'

/* @vue/component */
export default mixins(
  Toggleable,
  Transitionable,
  VSheet
).extend({
  name: 'v-alert',

  props: {
    closeIcon: {
      type: String,
      default: '$vuetify.icons.cancel'
    },
    dismissible: Boolean,
    icon: String,
    outline: Boolean,
    type: {
      type: String,
      validator (val: string) {
        return [
          'info',
          'error',
          'success',
          'warning'
        ].includes(val)
      }
    }
  },

  computed: {
    classes (): object {
      return {
        ...VSheet.options.computed.classes.call(this),
        'v-alert--outline': this.outline
      }
    },
    computedColor (): string {
      return this.color || this.type
    },
    computedIcon (): string | void {
      if (this.icon || !this.type) return this.icon

      switch (this.type) {
        case 'info': return '$vuetify.icons.info'
        case 'error': return '$vuetify.icons.error'
        case 'success': return '$vuetify.icons.success'
        case 'warning': return '$vuetify.icons.warning'
      }
    },
    isDark () {
      if (this.light) return false

      return this.computedColor
        ? true
        : Themeable.options.computed.isDark.call(this)
    }
  },

  methods: {
    genIcon () {
      if (!this.computedIcon) return null

      return this.$createElement('div', {
        staticClass: 'v-alert__action'
      }, [this.$createElement(VIcon, this.computedIcon)])
    },
    genContent () {
      return this.$createElement('div', this.$slots.default)
    },
    genDismissible () {
      if (!this.dismissible) return null

      const btn = this.$createElement(VBtn, {
        props: {
          icon: true
        },
        attrs: {
          'aria-label': 'Close'
        },
        on: {
          click: (e: Event) => {
            this.isActive = false
            this.$emit('click:close', e)
          }
        }
      }, [this.$createElement(VIcon, this.closeIcon)])

      return this.$createElement('div', {
        staticClass: 'v-alert__dismissible'
      }, [btn])
    }
  },

  render (h): VNode {
    const children = [
      this.genIcon(),
      this.genContent(),
      this.genDismissible()
    ]

    const setColor = this.outline ? this.setTextColor : this.setBackgroundColor
    const alert = h('div', setColor(this.computedColor, {
      staticClass: 'v-alert',
      class: this.classes,
      style: this.styles,
      attrs: {
        role: 'alert'
      },
      directives: [{
        name: 'show',
        value: this.isActive
      }],
      on: this.$listeners
    }), children)

    if (!this.transition) return alert

    return h('transition', {
      props: {
        name: this.transition,
        origin: this.origin,
        mode: this.mode
      }
    }, [alert])
  }
})
