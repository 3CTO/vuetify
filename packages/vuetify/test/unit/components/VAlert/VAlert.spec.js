import { test } from '@/test'
import VAlert from '@/components/VAlert'

test('VAlert.vue', ({ mount }) => {
  it('should be closed by default', () => {
    const wrapper = mount(VAlert)

    expect(wrapper.vm.isActive).toBe(false)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a close icon', () => {
    const wrapper = mount(VAlert, {
      propsData: { dismissible: true }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with transition', async () => {
    const wrapper = mount(VAlert, {
      propsData: { transition: 'foo' }
    })

    wrapper.setProps({ value: true })
    await wrapper.vm.$nextTick()
    expect(wrapper.hasClass('foo-enter')).toBe(true)
  })

  it('should render component with outline prop', () => {
    const wrapper = mount(VAlert, {
      propsData: { outline: true }
    })

    expect(wrapper.hasClass('v-alert--outline')).toBe(true)
  })

  it('should be dismissible', async () => {
    const wrapper = mount(VAlert, {
      propsData: {
        value: true,
        dismissible: true
      }
    })

    const btn = wrapper.first('.v-alert__dismissible .v-btn')

    const input = jest.fn(value => wrapper.setProps({ value }))
    wrapper.vm.$on('input', input)

    btn.trigger('click')

    await wrapper.vm.$nextTick()

    expect(input).toBeCalledWith(false)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a custom icon', () => {
    const wrapper = mount(VAlert, {
      propsData: {
        value: true,
        icon: 'list'
      }
    })

    const icon = wrapper.first('.v-icon')

    expect(icon.text()).toBe('list')
  })

  it('should have no icon', () => {
    const wrapper = mount(VAlert)

    expect(wrapper.contains('.v-icon')).toBe(false)
  })

  it('should display contextual colors by type', async () => {
    const wrapper = mount(VAlert, {
      propsData: {
        type: 'error'
      }
    })

    expect(wrapper.vm.computedColor).toBe('error')

    wrapper.setProps({ 'type': 'success' })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.computedColor).toBe('success')

    wrapper.setProps({ 'type': 'warning' })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.computedColor).toBe('warning')

    wrapper.setProps({ 'type': 'info' })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.computedColor).toBe('info')
  })

  it('should allow overriding color for contextual alert', () => {
    const wrapper = mount(VAlert, {
      propsData: {
        type: 'error',
        color: 'primary'
      }
    })

    expect(wrapper.vm.computedColor).toBe('primary')
  })

  it('should allow overriding icon for contextual alert', () => {
    const wrapper = mount(VAlert, {
      propsData: {
        type: 'error',
        icon: 'block'
      }
    })

    const icon = wrapper.first('.v-icon')

    expect(icon.text()).toBe('block')
  })

  it('should return a custom isDark', () => {
    const wrapper = mount(VAlert)

    expect(wrapper.vm.isDark).toBe(false)

    wrapper.setProps({ type: 'info' })

    expect(wrapper.vm.isDark).toBe(true)

    wrapper.setProps({
      color: 'blue' ,
      type: undefined
    })

    expect(wrapper.vm.isDark).toBe(true)

    wrapper.setProps({
      color: undefined,
      light: true
    })

    expect(wrapper.vm.isDark).toBe(false)
  })
})
