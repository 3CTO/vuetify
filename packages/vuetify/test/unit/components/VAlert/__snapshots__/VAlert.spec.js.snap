// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VAlert.vue should be closed by default 1`] = `

<div role="alert"
     class="v-alert v-sheet theme--light"
     style="display: none;"
>
  <div>
  </div>
</div>

`;

exports[`VAlert.vue should be dismissible 1`] = `

<div role="alert"
     class="v-alert v-sheet theme--light"
     style="display: none;"
>
  <div>
  </div>
  <div class="v-alert__dismissible">
    <button type="button"
            class="v-btn v-btn--icon theme--light"
            aria-label="Close"
    >
      <div class="v-btn__content">
        <i aria-hidden="true"
           class="v-icon material-icons theme--light"
        >
          cancel
        </i>
      </div>
    </button>
  </div>
</div>

`;

exports[`VAlert.vue should have a close icon 1`] = `

<div role="alert"
     class="v-alert v-sheet theme--light"
     style="display: none;"
>
  <div>
  </div>
  <div class="v-alert__dismissible">
    <button type="button"
            class="v-btn v-btn--icon theme--light"
            aria-label="Close"
    >
      <div class="v-btn__content">
        <i aria-hidden="true"
           class="v-icon material-icons theme--light"
        >
          cancel
        </i>
      </div>
    </button>
  </div>
</div>

`;
